File: src/2025-02-23-deepclaude-try/go/exercises/practice/dominoes/dominoes.go
package dominoes

type Domino [2]int

func MakeChain(input []Domino) ([]Domino, bool) {
	if len(input) == 0 {
		return []Domino{}, true
	}
	if len(input) == 1 {
		d := input[0]
		if d[0] == d[1] {
			return []Domino{d}, true
		}
		return nil, false
	}

	// Try each domino as first in both orientations
	for i := 0; i < len(input); i++ {
		// Original orientation
		used := make([]bool, len(input))
		used[i] = true
		start := input[i][0]
		end := input[i][1]
		if chain, ok := backtrack(input, used, []Domino{input[i]}, start, end); ok {
			return chain, true
		}
		// Flipped orientation
		used = make([]bool, len(input))
		used[i] = true
		flipped := Domino{input[i][1], input[i][0]}
		start = flipped[0]
		end = flipped[1]
		if chain, ok := backtrack(input, used, []Domino{flipped}, start, end); ok {
			return chain, true
		}
	}
	return nil, false
}

func backtrack(input []Domino, used []bool, chain []Domino, start, end int) ([]Domino, bool) {
	if len(chain) == len(input) {
		if start == end {
			return chain, true
		}
		return nil, false
	}

	for i := 0; i < len(input); i++ {
		if !used[i] {
			d := input[i]
			// Try original orientation
			if d[0] == end {
				newUsed := make([]bool, len(used))
				copy(newUsed, used)
				newUsed[i] = true
				newChain := append(chain, d)
				if result, ok := backtrack(input, newUsed, newChain, start, d[1]); ok {
					return result, true
				}
			}
			// Try flipped orientation
			if d[1] == end {
				newUsed := make([]bool, len(used))
				copy(newUsed, used)
				newUsed[i] = true
				flipped := Domino{d[1], d[0]}
				newChain := append(chain, flipped)
				if result, ok := backtrack(input, newUsed, newChain, start, flipped[1]); ok {
					return result, true
				}
			}
		}
	}

	return nil, false
}

File: src/2025-03-16-12-54-49--a-deepclaude-full-architect-python-0316-2/python/exercises/practice/dominoes/dominoes.py
def can_chain(dominoes):
    """
    Determine if a set of dominoes can form a chain where the first and last numbers match.
    
    Args:
        dominoes: List of domino stones represented as [a, b] where a and b are the numbers on each half.
    
    Returns:
        A correctly ordered list of dominoes forming a chain, or None if not possible.
    """
    if not dominoes:
        return []
    
    # For a single domino, it forms a valid chain only if both ends are the same
    if len(dominoes) == 1:
        a, b = dominoes[0]
        return dominoes if a == b else None
    
    # Build an adjacency list to represent possible connections
    graph = {}
    for i, (a, b) in enumerate(dominoes):
        if a not in graph:
            graph[a] = []
        if b not in graph:
            graph[b] = []
        graph[a].append((i, b))  # From a we can go to b using domino i
        graph[b].append((i, a))  # From b we can go to a using domino i
    
    # For a valid chain to exist, all vertices must have even degree
    # (except possibly two vertices, which would be the start and end)
    for value in graph.values():
        if len(value) % 2 != 0:
            # If there are more than 2 odd-degree vertices, no valid chain exists
            odd_vertices = sum(1 for v in graph.values() if len(v) % 2 != 0)
            if odd_vertices > 2:
                return None
    
    def find_chain(start, path=None, used=None):
        """Recursively find a valid domino chain starting from 'start'."""
        if path is None:
            path = []
        if used is None:
            used = set()
        
        # If we've used all dominoes, check if we can close the chain
        if len(used) == len(dominoes):
            return path if path[0][0] == path[-1][1] else None
        
        # Try each possible next domino
        for i, end in graph[start]:
            if i not in used:
                used.add(i)
                a, b = dominoes[i]
                # Add the domino in the correct orientation
                if a == start:
                    path.append([a, b])
                    result = find_chain(b, path, used)
                else:
                    path.append([b, a])
                    result = find_chain(a, path, used)
                
                if result:
                    return result
                
                # Backtrack
                path.pop()
                used.remove(i)
        
        return None
    
    # Try starting with each domino in both orientations
    for i, (a, b) in enumerate(dominoes):
        used = {i}
        
        # Try starting with [a, b]
        chain = find_chain(b, [[a, b]], used)
        if chain:
            return chain
        
        # Try starting with [b, a]
        chain = find_chain(a, [[b, a]], used)
        if chain:
            return chain
    
    return None
